File: java/tsfile/src/main/java/org/apache/tsfile/read/query/dataset/TableResultSet.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.tsfile.read.query.dataset;

import org.apache.tsfile.enums.TSDataType;
import org.apache.tsfile.read.common.block.TsBlock;
import org.apache.tsfile.read.reader.block.TsBlockReader;
import org.apache.tsfile.utils.DateUtils;
import org.apache.tsfile.write.record.TSRecord;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.time.LocalDate;
import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;

public class TableResultSet extends AbstractResultSet {

  private static final Logger LOG = LoggerFactory.getLogger(TableResultSet.class);

  private final TsBlockReader tsBlockReader;
  private TsBlock currentTsBlock;
  private int currentTsBlockIndex;
  private final List<String> columnNameList;
  private final List<TSDataType> dataTypes;
  private final String tableName;

  public TableResultSet(
      TsBlockReader tsBlockReader,
      List<String> columnNameList,
      List<TSDataType> dataTypeList,
      String tableName) {
    super(columnNameList, dataTypeList);
    this.columnNameList = columnNameList;
    this.dataTypes = dataTypeList;
    this.tsBlockReader = tsBlockReader;
    this.tableName = tableName;
  }

  @Override
  public boolean next() throws IOException {
    while ((currentTsBlock == null || currentTsBlockIndex >= currentTsBlock.getPositionCount() - 1)
        && tsBlockReader.hasNext()) {
      currentTsBlock = tsBlockReader.next();
      currentTsBlockIndex = -1;
    }
    currentTsBlockIndex++;
    return currentTsBlock != null && currentTsBlockIndex < currentTsBlock.getPositionCount();
  }

  @Override
  public int getInt(int columnIndex) {
    // -2 because the columnIndex starts from 1 and the first column is fixed as the time column
    return currentTsBlock.getValueColumns()[columnIndex - 2].getInt(currentTsBlockIndex);
  }

  @Override
  public long getLong(int columnIndex) {
    return columnIndex == 1
        ? currentTsBlock.getTimeByIndex(currentTsBlockIndex)
        // -2 because the columnIndex starts from 1 and the first column is fixed as the time column
        : currentTsBlock.getValueColumns()[columnIndex - 2].getLong(currentTsBlockIndex);
  }

  @Override
  public double getDouble(int columnIndex) {
    // -2 because the columnIndex starts from 1 and the first column is fixed as the time column
    return currentTsBlock.getValueColumns()[columnIndex - 2].getDouble(currentTsBlockIndex);
  }

  @Override
  public float getFloat(int columnIndex) {
    // -2 because the columnIndex starts from 1 and the first column is fixed as the time column
    return currentTsBlock.getValueColumns()[columnIndex - 2].getFloat(currentTsBlockIndex);
  }

  @Override
  public boolean getBoolean(int columnIndex) {
    // -2 because the columnIndex starts from 1 and the first column is fixed as the time column
    return currentTsBlock.getValueColumns()[columnIndex - 2].getBoolean(currentTsBlockIndex);
  }

  @Override
  public LocalDate getDate(int columnIndex) {
    // -2 because the columnIndex starts from 1 and the first column is fixed as the time column
    return DateUtils.parseIntToLocalDate(
        currentTsBlock.getValueColumns()[columnIndex - 2].getInt(currentTsBlockIndex));
  }

  @Override
  public byte[] getBinary(int columnIndex) {
    // -2 because the columnIndex starts from 1 and the first column is fixed as the time column
    return currentTsBlock
        .getValueColumns()[columnIndex - 2]
        .getBinary(currentTsBlockIndex)
        .getValues();
  }

  @Override
  public String getString(int columnIndex) {
    // -2 because the columnIndex starts from 1 and the first column is fixed as the time column
    return currentTsBlock
        .getValueColumns()[columnIndex - 2]
        .getBinary(currentTsBlockIndex)
        .toString();
  }

  @Override
  public boolean isNull(int columnIndex) {
    // -2 because the columnIndex starts from 1 and the first column is fixed as the time column
    return currentTsBlock.getValueColumns()[columnIndex - 2].isNull(currentTsBlockIndex);
  }

  @Override
  public void close() {
    if (tsBlockReader == null) {
      return;
    }
    try {
      tsBlockReader.close();
    } catch (Exception e) {
      LOG.error("Failed to close tsBlockReader");
    }
  }

  @Override
  public Iterator<TSRecord> iterator() {
    return new RecordIterator();
  }

  private class RecordIterator implements Iterator<TSRecord> {

    private TSRecord cachedRecord = null;
    private boolean exhausted = false;

    @Override
    public boolean hasNext() {
      if (cachedRecord != null) {
        return true;
      }
      if (exhausted) {
        return false;
      }

      try {
        return cacheNextRecord();
      } catch (IOException e) {
        throw new NoSuchElementException(e.toString());
      }
    }

    private boolean cacheNextRecord() throws IOException {
      boolean next = TableResultSet.this.next();
      if (!next) {
        exhausted = true;
        return false;
      }
      cachedRecord = new TSRecord(tableName, getLong("Time"));
      for (int i = 0; i < dataTypes.size(); i++) {
        TSDataType dataType = dataTypes.get(i);
        String columnName = columnNameList.get(i);
        if (isNull(columnName)) {
          cachedRecord.dataPointList.add(null);
          continue;
        }
        switch (dataType) {
          case INT32:
          case DATE:
            cachedRecord.addPoint(columnName, getInt(columnName));
            break;
          case INT64:
          case TIMESTAMP:
            cachedRecord.addPoint(columnName, getLong(columnName));
            break;
          case FLOAT:
            cachedRecord.addPoint(columnName, getFloat(columnName));
            break;
          case DOUBLE:
            cachedRecord.addPoint(columnName, getDouble(columnName));
            break;
          case STRING:
          case TEXT:
            cachedRecord.addPoint(columnName, getString(columnName));
            break;
          case BLOB:
            cachedRecord.addPoint(columnName, getBinary(columnName));
            break;
          case BOOLEAN:
            cachedRecord.addPoint(columnName, getBoolean(columnName));
            break;
          case VECTOR:
          case UNKNOWN:
          default:
            break;
        }
      }
      return true;
    }

    @Override
    public TSRecord next() {
      if (!hasNext()) {
        throw new NoSuchElementException();
      }
      TSRecord ret = cachedRecord;
      cachedRecord = null;
      return ret;
    }
  }
}

File: cpp/src/reader/table_result_set.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * License); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "reader/table_result_set.h"

#include <utils/storage_utils.h>

namespace storage {
void TableResultSet::init() {
    row_record_ = new RowRecord(column_names_.size() + 1);
    pa_.reset();
    pa_.init(512, common::MOD_TSFILE_READER);
    index_lookup_.reserve(column_names_.size() + 1);
    index_lookup_.insert({"time", 0});
    for (uint32_t i = 0; i < column_names_.size(); ++i) {
        index_lookup_.insert({column_names_[i], i + 1});
    }
    result_set_metadata_ =
        std::make_shared<ResultSetMetadata>(column_names_, data_types_);
}

TableResultSet::~TableResultSet() { close(); }

int TableResultSet::next(bool& has_next) {
    if (return_mode_ != RETURN_ROW) {
        return tsblock_reader_->has_next(has_next);
    }

    int ret = common::E_OK;

    // Advance past the row yielded by the previous next() call, if any.
    // Row iterator's next() advances all per-column offsets, so on the next
    // read the vectors point to the new row's data.
    if (row_ready_) {
        row_iterator_->next();
        row_ready_ = false;
        row_materialized_ = false;
    }

    // Find the next non-empty TsBlock.
    while (row_iterator_ == nullptr || !row_iterator_->has_next()) {
        if (RET_FAIL(tsblock_reader_->has_next(has_next))) {
            return ret;
        }

        if (!has_next) {
            if (row_iterator_) {
                delete row_iterator_;
                row_iterator_ = nullptr;
            }
            break;
        }

        if (RET_FAIL(tsblock_reader_->next(tsblock_))) {
            break;
        }
        if (row_iterator_) {
            delete row_iterator_;
            row_iterator_ = nullptr;
        }

        row_iterator_ = new common::RowIterator(tsblock_);
    }
    if (row_iterator_ == nullptr || !row_iterator_->has_next()) {
        has_next = false;
        return ret;
    }

    // A row is now available at row_iterator_'s current row_id_; the per-
    // column vector offsets are pointing at that row's data.  We do NOT
    // populate row_record_ here — typed accessors read straight from the
    // vectors, and get_row_record() lazily materializes on demand.
    has_next = true;
    row_ready_ = true;
    return ret;
}

void TableResultSet::materialize_current_row() {
    uint32_t len = 0;
    bool null = false;
    row_record_->reset();
    for (uint32_t i = 0; i < row_iterator_->get_column_count(); ++i) {
        const auto value = row_iterator_->read(i, &len, &null);
        if (!null) {
            row_record_->get_field(i)->set_value(
                row_iterator_->get_data_type(i), value, len, pa_);
        }
    }
}

bool TableResultSet::is_null(const std::string& column_name) {
    auto iter = index_lookup_.find(column_name);
    if (iter == index_lookup_.end()) {
        return true;
    } else {
        return is_null(iter->second + 1);
    }
}

bool TableResultSet::is_null(uint32_t column_index) {
    ASSERT(1 <= column_index && column_index <= row_record_->get_col_num());
    if (!row_ready_) return true;
    return row_iterator_->is_null_at(column_index - 1);
}

// Direct buffer access — skips Vector::read's virtual dispatch.  Caller is
// expected to have checked is_null() (we still null-guard for safety).
// For fixed-width primitives the vector keeps its value buffer in
// values_ and tracks the current row's byte offset in offset_; the
// element at the active row is simply *(T*)(values_.get_data() + offset_).
// The ASSERT enforces strict typed access: the requested C++ type must match
// the column's physical storage width (DATE is int32, not int64).  On a
// mismatch it fires in debug instead of silently splicing the adjacent cell's
// bytes into the result.
#define TSFILE_FAST_PRIMITIVE_READ(TYPE, DFLT)                         \
    if (!row_ready_) return DFLT;                                      \
    common::Vector* vec = row_iterator_->get_vector(column_index - 1); \
    ASSERT(common::TypeMatch<TYPE>(vec->get_vector_type()));           \
    if (vec->has_null() && vec->is_null(row_iterator_->get_row_id()))  \
        return DFLT;                                                   \
    return *reinterpret_cast<TYPE*>(vec->get_value_data().get_data() + \
                                    vec->get_offset())

bool TableResultSet::get_bool_at(uint32_t column_index) {
    TSFILE_FAST_PRIMITIVE_READ(bool, false);
}

int32_t TableResultSet::get_int32_at(uint32_t column_index) {
    TSFILE_FAST_PRIMITIVE_READ(int32_t, 0);
}

int64_t TableResultSet::get_int64_at(uint32_t column_index) {
    TSFILE_FAST_PRIMITIVE_READ(int64_t, 0);
}

float TableResultSet::get_float_at(uint32_t column_index) {
    TSFILE_FAST_PRIMITIVE_READ(float, 0.0f);
}

double TableResultSet::get_double_at(uint32_t column_index) {
    TSFILE_FAST_PRIMITIVE_READ(double, 0.0);
}

#undef TSFILE_FAST_PRIMITIVE_READ

RowRecord* TableResultSet::get_row_record() {
    if (row_ready_ && !row_materialized_) {
        materialize_current_row();
        row_materialized_ = true;
    }
    return row_record_;
}

std::shared_ptr<ResultSetMetadata> TableResultSet::get_metadata() {
    return result_set_metadata_;
}

int TableResultSet::get_next_tsblock(common::TsBlock*& block) {
    int ret = common::E_OK;
    block = nullptr;

    if (return_mode_ == RETURN_ROW) {
        return common::E_INVALID_ARG;
    }

    bool has_next = false;
    if (RET_FAIL(tsblock_reader_->has_next(has_next))) {
        return ret;
    }

    if (!has_next) {
        return common::E_NO_MORE_DATA;
    }

    if (RET_FAIL(tsblock_reader_->next(tsblock_))) {
        return ret;
    }

    if (tsblock_ == nullptr) {
        return common::E_NO_MORE_DATA;
    }

    block = tsblock_;
    return common::E_OK;
}

void TableResultSet::close() {
    if (closed_) {
        return;
    }
    closed_ = true;
    if (tsblock_reader_) {
        tsblock_reader_->close();
    }
    pa_.destroy();
    if (row_record_) {
        delete row_record_;
        row_record_ = nullptr;
    }
    if (row_iterator_) {
        delete row_iterator_;
        row_iterator_ = nullptr;
    }
}

}  // namespace storage
